haskell/test/SMT/Model/BaseDAO/Types.hs

```diff
@@ -0,0 +1,263 @@
+-- SPDX-FileCopyrightText: 2021 TQ Tezos
+-- SPDX-License-Identifier: LicenseRef-MIT-TQ
+
+{-# OPTIONS_GHC -Wno-orphans #-}
+{-# OPTIONS_GHC -Wno-missing-signatures #-}
+
+module SMT.Model.BaseDAO.Types
+  ( ModelT (..)
+  , ModelState (..)
+  , SimpleContractType (..)
+  , SimpleFA2Contract (..)
+  , OtherContract (..)
+  , SimpleOperation (..)
+  , runModelT
+  , getStore
+  , getConfig
+  , modifyStore
+  , execOperation
+
+  , ContractType (..)
+  , ModelCall (..)
+  , ModelSource (..)
+  , ModelError (..)
+  , contractErrorToModelError
+  ) where
+
+import Universum hiding (show)
+
+import Control.Monad.Except (MonadError)
+import qualified Data.Map as Map
+import Fmt
+import Text.Show (show)
+
+import Michelson.Text
+import Lorentz hiding (not, cast, get)
+import qualified Lorentz.Contracts.Spec.FA2Interface as FA2
+import Tezos.Address (ContractHash(..))
+import Tezos.Core
+
+import Ligo.BaseDAO.Types
+
+-- | Transformer used in haskell implementation of BaseDAO
+newtype ModelT a = ModelT
+  { unModelT :: (ExceptT ModelError $ State ModelState) a
+  } deriving newtype (Functor, Applicative, Monad, MonadState ModelState, MonadError ModelError)
+
+-- | A type that keep track of blockchain state
+data ModelState = ModelState
+  { msFullStorage :: FullStorage
+  , msMutez :: Mutez
+  -- ^ Dao balance
+
+  , msLevel :: Natural
+  , msChainId :: ChainId
+  , msSelfAddress :: Address
+  , msGovernanceTokenAddress :: Address
+  , msContracts :: Map Address SimpleContractType
```

Comment:
This map only ever contains 2 contracts, right?

If so, would it not be easier to replace it with 2 fields (or a record with 2 fields) for each contract? That way, we wouldn't need `Map.lookup`, `Map.insertt`, `fromMaybe (error "Governance contract does not exist")`, etc

Reply:
I'm going to refer to my comment here: https://github.com/tqtezos/baseDAO/pull/282#discussion_r666682832

Basically, the current generator only cares about this 2 addresses. In the future, we make keep track of other contract storage. 

The Haskell BaseDAO work with a `Map` since it acts as a mock for all contract in a blockchain.

Reply:
Even when (if?) we have more contracts to keep track of, it'll still be a fixed number of contracts, right? 

Personally, I tend to go with the YAGNI principle, i.e., don't complicate the code _now_ in favor of something that _might_ (or might not) happen in the future. My predictions are often wrong :sweat_smile: But I'll leave this decision up to you.